packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+
+    auto packageNamespace =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(package.name.mangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(importedPackages));
+
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    rootKlass->rhs.emplace_back(move(packageNamespace));
+    return file;
+}
+
+ast::ParsedFile rewritePackagedFile(core::Context ctx, ast::ParsedFile file, core::NameRef packageMangledName) {
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    auto moduleWrapper =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(packageMangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(rootKlass->rhs));
+    rootKlass->rhs.clear();
+    rootKlass->rhs.emplace_back(move(moduleWrapper));
+    return file;
+}
+
+// We can't run packages without having all package ASTs. Assert that they are all present.
+bool checkContainsAllPackages(const core::GlobalState &gs, const vector<ast::ParsedFile> &files) {
+    UnorderedSet<core::FileRef> allPackages;
+    for (u4 i = 1; i < gs.filesUsed(); i++) {
+        core::FileRef fref(i);
+        if (fref.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.insert(fref);
+        }
+    }
+
+    for (const auto &f : files) {
+        if (f.file.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.erase(f.file);
+        }
+    }
+    return allPackages.empty();
+}
+
+} // namespace
+
+vector<ast::ParsedFile> Packager::run(core::GlobalState &gs, WorkerPool &workers, vector<ast::ParsedFile> files) {
+    Timer timeit(gs.tracer(), "packager");
+    // Ensure files are in canonical order.
+    fast_sort(files, [](const auto &a, const auto &b) -> bool { return a.file < b.file; });
+
+    // Step 1: Find packages and determine their imports/exports.
+    UnorderedMap<core::FileRef, shared_ptr<PackageInfo>> packageInfoByFile;
+    UnorderedMap<core::NameRef, shared_ptr<PackageInfo>> packageInfoByMangledName;
+    vector<shared_ptr<PackageInfo>> packages;
+    {
+        Timer timeit(gs.tracer(), "packager.findPackages");
+        core::UnfreezeNameTable unfreeze(gs);
+        for (auto &file : files) {
+            if (FileOps::getFileName(file.file.data(gs).path()) == PACKAGE_FILE_NAME) {
+                file.file.data(gs).sourceType = core::File::Type::Package;
+                core::MutableContext ctx(gs, core::Symbols::root(), file.file);
+                auto pkg = getPackageInfo(ctx, file);
+                // getPackageInfo emits an error if the package is malformed.
+                if (pkg != nullptr) {
```

Comment:
```suggestion
                // getPackageInfo emits an error if the package is malformed.
                if (auto pkg = getPackageInfo(ctx, file)) {
```

This makes it impossible to accidentally use `pkg` in a static scope where it's guaranteed to be `nullptr`.

Reply:
I see you using this pattern a lot in this file actually. Are you doing that intentionally? I think that a quick grep for `!= nullptr` in this file will show you a bunch of other places where you could benefit from this pattern.